### ISSUE TITLE
vi mode control-f

### ISSUE THREAD

**Issue by bobbobbio** [fish-shell/fish-shell#2255]
I was curious why vi mode changes the behavior of control-f, control-b, control-a, control-e in insert mode.  This key bindings don't really do anything in vi insert mode I don't think, so why not retain existing behavior? I can sort of understand that in vi mode you would want to do these actions with in command mode with more vi-like commands, but then again that which they have been remapped to is also possible in vi command mode, so why not just leave them be?

One aside related to this though is that with control-f remapped there is no way to take the current completion in insert mode without resorting to the arrow keys.  Maybe there needs to be some other way to take the current completion in insert mode, (this seems especially salient in vi mode since a huge goal of the vi key bindings is not having to have your hands leave the home-row)

**Comment by pickfire** [comment]
I hope you can clarify your request to show others what you mean, such as:
```
^f -> home row
```

**Comment by faho** [comment]
In vim, ^f scrolls a page down, ^b scrolls up, ^e scrolls down a line and ^a goes back a character or increments the number under the cursor - is that really what it does? - none of which seem particularly useful to a shell except maybe ^e which could be made to do history-search-forward.

There are three things we could do here: 

- Leave it unbound and up to the user to do something with it

- Bind it to the same thing as the default bindings

- Find a vi-esque thing to bind to it

I like the second option since it lessens friction when switching from the default bindings to the vi-ones.

**Comment by jonashaag** [comment]
<kbd>Control</kbd>+<kbd>F</kbd> still doesn't accept the autocompletion if the cursor's at the end of line.

Default key bindings:

```
~> rm -r [some/sub/directory]
^F
~> rm -r some/sub/directory
<Return>
```

Vim key bindings:

```
~> rm -r [some/sub/directory]
^F
~> rm -r some[/sub/directory]
<Return>
```

My suggestion is to never change the way autocompletion keybindings work to save users from executing unexpected commands, like `rm` with unfinished arguments...

**Comment by faho** [comment]
><kbd>Control</kbd>+<kbd>F</kbd> still doesn't accept the autocompletion if the cursor's at the end of line.

Sorry, @pickfire confused me in #2472. Anyway, it should have been removed from the documentation, and it seems like some people like the current way.

So, @pickfire: Mind participating here? Why should \cf act differently in vi-mode than in emacs-mode? It doesn't act that way in vi or vim-proper.

**Comment by pickfire** [comment]
Hi @faho, I am here now. Sorry for the late reply.

I think I prefer it to be \cx in vi-mode rather than \cf because vim uses \cx to enter completion mode by default. In My Humble Opinion, I am not saying that I doesn't want \cf to be `end-of-line` and instead I just hope that it can use the default binding for completion.

And for \cf, I doesn't mind what it is. If it does the same as `forward-char`, I will just use it, any command for that will do for me.

**Comment by faho** [comment]
So, here's what I've done for now:

- Made \cx do "end-of-line", which will accept suggestions (this has been added to the documentation)

- Made \cb do "backward-word" again (and kept \cf as forward-word) like the documentation says

I'm not sure that two bindings so close to each other should do the same thing and I don't like re-defining bindings and breaking muscle-memory for those used to the old ways without a good reason.

So I'm going to keep it this way unless anyone objects.

**Comment by jonashaag** [comment]
I'm still opposed to changing <kbd>Control</kdb>+<kdb>F<k/db> behaviour between the default and vim keybindings for the reason stated in https://github.com/fish-shell/fish-shell/issues/2255#issuecomment-148385160.

**Comment by pickfire** [comment]
@jonashaag The default vim binding doesn't use \cf as completion but \cx as completion.

In a nutshell, [\cf execute `forward-word`](https://github.com/fish-shell/fish-shell/blob/master/share/functions/fish_vi_key_bindings.fish#L168) and [\cx execute `end-of-line`](https://github.com/fish-shell/fish-shell/blob/master/share/functions/fish_vi_key_bindings.fish#L166).

**Comment by jonashaag** [comment]
@pickfire Yeah but it's `forward-char

**Comment by pickfire** [comment]
@jonashaag I am not saying that I do not want to change it but if you change \cf to `forward-word`, what will \cx do? And how do I execute `forward-word` (\cw)?

**Comment by jonashaag** [comment]
> @jonashaag I am not saying that I do not want to change it but if you change \cf to forward-word, what will \cx do? And how do I execute forward-word (\cw)?

- <kbd>Control</kbd>+<kbd>F</kbd>: `forward-char` (as in default keybindings)
- <kbd>Control</kbd>+<kbd>X</kbd>: `end-of-line` (personally I don't need this but it doesn't hurt either)
- <kbd>Control</kbd>+<kbd>W</kbd>: delete word (as in default keybindings)
- <kbd>Alt</kbd>+<kbd>→</kbd>: `forward-word` (as in default keybindings)

**Comment by pickfire** [comment]
That's what I am saying. But I mean `\cf` as `forward-word`.

**Comment by jonashaag** [comment]
bump -- I'd still like to see a decision on this.

**Comment by pickfire** [comment]
Yeah, I hope someone can make a decision on this.

**Comment by faho** [comment]
I've not forgotten about this, the decision is just rather hard.

We have a desire to have as little delta between the vi-bindings and the emacs-bindings as possible - essentially whenever the vi-nature doesn't dictate something else, we should keep the binding - on the one side, and a desire to not just _change_ bindings left and right every release (I'm okay with _adding_ them) so we don't break muscle memory.

The function that was asked about (a way to accept suggestions without switching to normal/default mode or using the error keys) is also already possible with \cx. Also, just for the record, bash and zsh seem to insert "^F" on \cf (i.e. do nothing worthwhile).

**BUT**, and this is the argument that solves it for me, these bindings were added in ed5e5856844a68f0f5c6c0cc84ec791dec78c722, which could have been released in 2.1.1 at most, so it's only been there for two releases - though the 2.2.0 release notes mention vi-mode.

So I'm going to remove it from the code _and_ the docs.

**Comment by jonashaag** [comment]
:tada: Thank you!
